### PR TITLE
Fixes to indentation work

### LINKED ISSUE #151: Question about indentation check (L003)
This query reports lots of indentation issues, but I think the indentation is along the lines of what you're planning to enforce. Is this the intended behavior? (I think you mentioned recently that query formatting is still a work in progress.)

```
WITH
    audience_counts AS (
        SELECT
            user_id,
            list_id,
            COUNT(email_id) AS audience
        FROM
            lists_emails AS list_emails
        WHERE
            list_emails.active != 'D'
        GROUP BY
            user_id,
            list_id)

SELECT
    user_id,
    list_id,
    audience,
    CASE
        WHEN audience > 0 AND audience <= 200 THEN '< 200'
        WHEN audience > 200
            AND audience <= 2000 THEN '200 - 2,000'
        WHEN audience > 2000 AND audience <= 10000 THEN '2,000 - 10,000'
        WHEN audience > 10000
            AND audience <= 50000 THEN '10,000 - 50,000'
        WHEN audience > 50000 AND audience <= 500000 THEN '50,000 - 500,000'
        WHEN audience > 500000 THEN '> 500,000'
    END AS size_bucket
FROM
    audience_counts
JOIN
    gdpr_safe_users
USING
    (user_id)
```

Results:
```
$ sqlfluff lint test.sql
== [test.sql] FAIL
L:   2 | P:   5 | L003 | Indentation not consistent with line #1
L:   3 | P:   9 | L003 | Line over-indented compared to line #1
L:   4 | P:  13 | L003 | Line over-indented compared to line #1
L:   5 | P:  13 | L003 | Line over-indented compared to line #1
L:   6 | P:  13 | L003 | Line over-indented compared to line #1
L:   7 | P:   9 | L003 | Line over-indented compared to line #1
L:   8 | P:  13 | L003 | Line over-indented compared to line #1
L:   9 | P:   9 | L003 | Line over-indented compared to line #1
L:  10 | P:  13 | L003 | Line over-indented compared to line #1
L:  11 | P:   9 | L003 | Line over-indented compared to line #1
L:  12 | P:  13 | L003 | Line over-indented compared to line #1
L:  13 | P:  13 | L003 | Line over-indented compared to line #1
L:  20 | P:   9 | L003 | Indentation not consistent with line #18
L:  21 | P:   9 | L003 | Indentation not consistent with line #18
L:  22 | P:  13 | L003 | Indentation not consistent with line #18
L:  23 | P:   9 | L003 | Indentation not consistent with line #18
L:  24 | P:   9 | L003 | Indentation not consistent with line #18
L:  25 | P:  13 | L003 | Indentation not consistent with line #18
L:  26 | P:   9 | L003 | Indentation not consistent with line #18
L:  27 | P:   9 | L003 | Indentation not consistent with line #18
L:  30 | P:   5 | L003 | Indentation not consistent with line #28
L:  32 | P:   5 | L003 | Indentation not consistent with line #30
L:  34 | P:   5 | L003 | Indentation not consistent with line #32
```

After running `sqlfluff fix --rules L003` on the file, it looks like this:
```
WITH
audience_counts AS (
    SELECT
        user_id,
        list_id,
        COUNT(email_id) AS audience
    FROM
    lists_emails AS list_emails
    WHERE
    list_emails.active != 'D'
    GROUP BY
    user_id,
    list_id)

SELECT
    user_id,
    list_id,
    audience,
    CASE
    WHEN audience > 0 AND audience <= 200 THEN '< 200'
    WHEN audience > 200
    AND audience <= 2000 THEN '200 - 2,000'
    WHEN audience > 2000 AND audience <= 10000 THEN '2,000 - 10,000'
    WHEN audience > 10000
    AND audience <= 50000 THEN '10,000 - 50,000'
    WHEN audience > 50000 AND audience <= 500000 THEN '50,000 - 500,000'
    WHEN audience > 500000 THEN '> 500,000'
    END AS size_bucket
FROM
audience_counts
JOIN
gdpr_safe_users
USING
(user_id)
```

Specifically, I was thinking the lines after `FROM`, `WHERE`, `GROUP BY`, `CASE`, `JOIN`, and `USING` would be indented.

### PR DESCRIPTION
**Pull request by alanmcruickshank**
fixes #151 .

@barrywhart @mrshu - can you review?

I've ironed out some bugs in the indentation code, but also been thinking about the *correct* indentation for the sample query, and I'd like to suggest a few changes. I think the correct indentation should look like this:

```
WITH
    audience_counts AS (
        SELECT
            user_id,
            list_id,
            COUNT(email_id) AS audience
        FROM
            lists_emails AS list_emails
        WHERE
            list_emails.active != 'D'
        GROUP BY
            user_id,
            list_id)

SELECT
    user_id,
    list_id,
    audience,
    CASE
        WHEN audience > 0 AND audience <= 200 THEN '< 200'
        WHEN audience > 200
            AND audience <= 2000 THEN '200 - 2,000'
        WHEN audience > 2000 AND audience <= 10000 THEN '2,000 - 10,000'
        WHEN audience > 10000
            AND audience <= 50000 THEN '10,000 - 50,000'
        WHEN audience > 50000 AND audience <= 500000 THEN '50,000 - 500,000'
        WHEN audience > 500000 THEN '> 500,000'
    END AS size_bucket
FROM
    audience_counts
    JOIN
        gdpr_safe_users
        USING
            (user_id)
```

With the notable changes being in the `FROM` clause. I think the `JOIN` is technically part of the `FROM` and so should be indented from it. Likewise the `USING` clause is part of the `JOIN` and so should likewise be indented.

Do you both agree with that suggestion? If it's a blanket *yes* then I'll just merge it. If you don't agree, then I think this might be a case for adding a configurable flag, where this is the default setting, but it can be over-ridden. In that case I'm assuming it's the indentation of `JOIN` which is going to be the most controversial (heck, I don't even indent it like this at the moment) - so that seems like the obvious place.

On the other hand, I'd like sqlfluff to encourage people to do things *correctly* if at all possible, so I think if we agree what is *correct*, then we should make that the default.

Thoughts?

### CODE REVIEW

**Comment by barrywhart** [comment]
Ha ha, I am in the same situation as you -- your point about `JOIN` makes logical sense, but I have always treated `JOIN` as being at the outer level, so I will find it very difficult to change. I'm even more worried about the data scientists than myself -- it's _their_ SQL that I'm mainly thinking of for linting. 😬

File: src/sqlfluff/rules/std.py [hunk 1]
```diff
@@ -337,17 +337,24 @@ def _eval(self, segment, raw_stack, memory, **kwargs):
                 if this_line['indent_size'] != res[k]['indent_size']:
                     # Indents don't match even though balance is the same...
                     memory['problem_lines'].append(this_line_no)
+                    print(this_line_no, k, this_line['indent_balance'], res[k]['indent_balance'])
```
Comment (barrywhart) [inline src/sqlfluff/rules/std.py]: Did you mean to leave the `print()` statements here and below?

Comment (alanmcruickshank) [reply src/sqlfluff/rules/std.py]: Ha! Nope I'll pull that one out. Good catch. Debugging this one was hard 🤕

**Comment by alanmcruickshank** [comment]
@barrywhart - suggestion. Assuming the rest of the code changes in here look good. How about we merge this version, mostly to get 0.3.x out, and then add a new issue for making the indentation of `JOIN` configurable. It might be quite a tricky one to have because it would require conditions on the `Indent` and `Dedent` objects.

**Comment by barrywhart** [comment]
Sounds good#

**Comment by mrshu** [comment]
@alanmcruickshank I also gave the code a quick read and it looks good to me.

With regards to the the `JOIN` being part of the `FROM`, I do think you are technically correct but the reality is that they are being though of as "separate parts of query configuration", for lack of a better term. Similarly to what @barrywhart has mentioned above, it would most probably be very difficult to change, especially since them being on the outer level is implicitly part of things like the [Mozilla SQL Style Guide](https://docs.telemetry.mozilla.org/concepts/sql_style.html).

Your plan for getting this released sounds good though!

**Comment by alanmcruickshank** [comment]
Ok guys. @barrywhart @mrshu I've reverted it back to the traditional positioning of the `JOIN` clause. I've left a TODO in the code to remind us how to change it if we want to. I'll go ahead and merge this.

**Pull request merged**